test/test_lazy.py

```diff
@@ -0,0 +1,233 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from functools import partial
+from pytools.obj_array import make_obj_array, obj_array_vectorize_n_args
+import pyopencl as cl
+import pyopencl.tools as cl_tools
+import pyopencl.array as cla  # noqa
+import pyopencl.clmath as clmath  # noqa
+from meshmode.array_context import (  # noqa
+    PyOpenCLArrayContext,
+    PytatoPyOpenCLArrayContext
+)
+from arraycontext.container.traversal import freeze, thaw
+
+from meshmode.array_context import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests)
+
+import pytest  # noqa
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def op_test_data(ctx_factory):
+    cl_ctx = ctx_factory()
+    queue = cl.CommandQueue(cl_ctx)
+
+    eager_actx = PyOpenCLArrayContext(
+        queue,
+        allocator=cl_tools.MemoryPool(cl_tools.ImmediateAllocator(queue)))
+
+    lazy_actx = PytatoPyOpenCLArrayContext(
+        queue,
+        allocator=cl_tools.MemoryPool(cl_tools.ImmediateAllocator(queue)))
+
+    def get_discr(order):
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        mesh = generate_regular_rect_mesh(
+            a=(-0.5,)*2,
+            b=(0.5,)*2,
+            nelements_per_axis=(4,)*2,
+            boundary_tag_to_face={
+                "x": ["-x", "+x"],
+                "y": ["-y", "+y"]
+            })
+
+        from grudge.eager import EagerDGDiscretization
+        return EagerDGDiscretization(eager_actx, mesh, order=order)
+
+    return eager_actx, lazy_actx, get_discr
+
+
+def _rel_linf_error(discr, actual, expected):
+    from mirgecom.fluid import ConservedVars
+    if isinstance(actual, ConservedVars):
+        return _rel_linf_error(discr, actual.join(), expected)
+    if isinstance(expected, ConservedVars):
+        return _rel_linf_error(discr, actual, expected.join())
+    return np.max(obj_array_vectorize_n_args(
+        lambda a, b: discr.norm(a - b, np.inf) / discr.norm(b, np.inf),
+        actual, expected))
```

Comment:
Lambdas and `obj_array_vectorize_n_args` are tough for me.  I've read up on these - but please have patience if I mess up what this is doing.   What I _think_ it is doing is:

```
max_component( max_domain(local_component_error) / max_domain(local_component_value) )
```

If that's true, one of my remaining concerns is that the numerator and denominator are not spatially correlated.  Why not something like this:

```
max_component( max_domain (local_component_error / local_component_value) )
```

.. or, in words, the maximum relative error of the component?  This is a real question.  Then the only thing left to worry about is whether `local_component_value` is dinky, in which case we could address with something along the lines of:

```
max_component( max_domain(compare_values( value1, value2, tol)
```

.. with `compare_values` handling the floating point value comparison by some method we can look up - returning relative errors where appropriate and possible, otherwise a different, but representative comparison.

Reply:
> Lambdas and `obj_array_vectorize_n_args` are tough for me. I've read up on these - but please have patience if I mess up what this is doing.

Lambdas are just a shorthand for functions. The above is equivalent to:

```python
    def component_error(a, b):
        discr.norm(a - b, np.inf) / discr.norm(b, np.inf)

    return np.max(obj_array_vectorize_n_args(component_error, actual, expected))
```

`obj_array_vectorize` and `obj_array_vectorize_n_args` just take a function, apply it to each component of an object array (or multiple object arrays for the `_n_args` version) and spit back the results in an object array.

So it's equivalent to something like:

```python
    def component_error(a, b):
        discr.norm(a - b, np.inf) / discr.norm(b, np.inf)

    ncomponents = len(actual)
    comp_errors = np.empty(ncomponents, dtype=object)
    for i in range(ncomponents):
        comp_errors[i] = component_error(actual[i], expected[i])

    return np.max(comp_errors)
```

> What I _think_ it is doing is:
> 
> ```
> max_component( max_domain(local_component_error) / max_domain(local_component_value) )
> ```

Yup.

> If that's true, one of my remaining concerns is that the numerator and denominator are not spatially correlated. Why not something like this:
> 
> ```
> max_component( max_domain (local_component_error / local_component_value) )
> ```
> 
> .. or, in words, the maximum relative error of the component? This is a real question.

Here I'm kind of thinking about the denominator as if it was a nondimensionalization factor, i.e., it's just something that's representative of the overall scale of that component.

> Then the only thing left to worry about is whether `local_component_value` is dinky, in which case we could address with something along the lines of:
> 
> ```
> max_component( max_domain(compare_values( value1, value2, tol)
> ```
> 
> .. with `compare_values` handling the floating point value comparison by some method we can look up - returning relative errors where appropriate and possible, otherwise a different, but representative comparison.

This sounds fine to me, I just don't know offhand how to deal with the divide-by-zero etc. issues gracefully.

Reply:
> Lambdas are just a shorthand for functions. The above is equivalent to:
> 
> `obj_array_vectorize` and `obj_array_vectorize_n_args` just take a function, apply it to each component of an object array (or multiple object arrays for the `_n_args` version) and spit back the results in an object array.
> 

Thanks for that @majosm.  Very clear. I'll get used to it :)

> Here I'm kind of thinking about the denominator as if it was a nondimensionalization factor, i.e., it's just something that's representative of the overall scale of that component.
>

I do think that makes sense, but is likely to not yield what you want when the field values vary a great deal across the domain - which is not atypical of our simulations.
   
> > .. with `compare_values` handling the floating point value comparison by some method we can look up - returning relative errors where appropriate and possible, otherwise a different, but representative comparison.
> 
> This sounds fine to me, I just don't know offhand how to deal with the divide-by-zero etc. issues gracefully.

I like this:

https://floating-point-gui.de/errors/comparison/